registry/storage/tagstore.go

```diff
@@ -134,44 +153,118 @@ func (ts *tagStore) linkedBlobStore(ctx context.Context, tag string) *linkedBlob
 
 // Lookup recovers a list of tags which refer to this digest.  When a manifest is deleted by
 // digest, tag entries which point to it need to be recovered to avoid dangling tags.
-func (ts *tagStore) Lookup(ctx context.Context, desc distribution.Descriptor) ([]string, error) {
+func (ts *TagStore) Lookup(ctx context.Context, desc distribution.Descriptor) ([]string, error) {
 	allTags, err := ts.All(ctx)
 	switch err.(type) {
+	case nil:
+		break
 	case distribution.ErrRepositoryUnknown:
 		// This tag store has been initialized but not yet populated
 		break
-	case nil:
-		break
 	default:
 		return nil, err
 	}
 
-	var tags []string
-	for _, tag := range allTags {
-		tagLinkPathSpec := manifestTagCurrentPathSpec{
-			name: ts.repository.Named().Name(),
-			tag:  tag,
+	lookupErr := &atomicError{}
+
+	allTagsCount := len(allTags)
+	inputChan := make(chan string)
+	outputChan := make(chan string, allTagsCount)
+
+	waitGroup := sync.WaitGroup{}
+
+	limiter := make(chan struct{}, ts.lookupConcurrencyFactor)
+	acquire := func() {
+		limiter <- struct{}{}
+		waitGroup.Add(1)
+	}
+	release := func() {
+		<-limiter
+		waitGroup.Done()
+	}
+
+	go func() {
+		defer func() {
+			waitGroup.Wait()
+			close(outputChan)
+			close(limiter)
+		}()
+		for tag := range inputChan {
+			acquire()
+			go func(tag string) {
+				defer release()
+
+				// No need to lookup further on lookupErr
+				if lookupErr.Load() != nil {
+					return
+				}
+
+				tagLinkPathSpec := manifestTagCurrentPathSpec{
+					name: ts.repository.Named().Name(),
+					tag:  tag,
+				}
+
+				tagLinkPath, _ := pathFor(tagLinkPathSpec)
+				tagDigest, err := ts.blobStore.readlink(ctx, tagLinkPath)
+
+				if err != nil {
+					switch err.(type) {
```

Comment:
Use `errors.Is` / `errors.As` instead of type-asserting on the error value. See https://pkg.go.dev/errors for more details.